golang/cosmos/x/swingset/keeper/snapshotter.go

```diff
@@ -0,0 +1,421 @@
+		BlockHeight: snapshotter.activeSnapshot.height,
+		Request:     "retrieve",
+	}
+	out, err := snapshotter.blockingSend(action)
+
+	if err != nil {
+		return err
+	}
+
+	var exportDir string
+	err = json.Unmarshal([]byte(out), &exportDir)
+	if err != nil {
+		return err
+	}
+
+	defer os.RemoveAll(exportDir)
+
+	rawManifest, err := os.ReadFile(filepath.Join(exportDir, ExportManifestFilename))
+	if err != nil {
+		return err
+	}
+
+	var manifest exportManifest
+	err = json.Unmarshal(rawManifest, &manifest)
+	if err != nil {
+		return err
+	}
+
+	if manifest.BlockHeight != height {
+		return fmt.Errorf("snapshot manifest blockHeight (%d) doesn't match (%d)", manifest.BlockHeight, height)
+	}
+
+	fileToPayload := func(name string, filename string) error {
+		payload := types.ExtensionSnapshotterArtifactPayload{Name: name}
+
+		payload.Data, err = os.ReadFile(filepath.Join(exportDir, filename))
+		if err != nil {
+			return err
+		}
+
+		payloadBytes, err := payload.Marshal()
+		if err != nil {
+			return err
+		}
+
+		err = payloadWriter(payloadBytes)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	if manifest.Data != "" {
+		err = fileToPayload(UntrustedExportDataArtifactName, manifest.Data)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, artifactInfo := range manifest.Artifacts {
+		name := artifactInfo[0]
+		filename := artifactInfo[1]
+		if name == UntrustedExportDataArtifactName {
+			return fmt.Errorf("unexpected artifact name %s", name)
+		}
+		err = fileToPayload(name, filename)
+		if err != nil {
+			return err
+		}
+	}
+
+	snapshotter.activeSnapshot.retrieved = true
+
+	snapshotter.logger.Info("fully retrieved snapshot from", "exportDir", exportDir)
+
+	return nil
+}
+
+// RestoreExtension restores an extension state snapshot,
+// the payload reader returns `io.EOF` when it reaches the extension boundaries.
+func (snapshotter *SwingsetSnapshotter) RestoreExtension(height uint64, format uint32, payloadReader snapshots.ExtensionPayloadReader) error {
+	if format != SnapshotFormat {
+		return snapshots.ErrUnknownFormat
+	}
+
+	ctx := snapshotter.app.NewUncachedContext(false, tmproto.Header{Height: int64(height)})
+
+	exportDir, err := os.MkdirTemp("", fmt.Sprintf("agd-state-sync-restore-%d-*", height))
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(exportDir)
+
+	manifest := exportManifest{
+		BlockHeight: height,
+		Data:        ExportDataFilename,
+	}
+
+	exportDataFile, err := os.OpenFile(filepath.Join(exportDir, ExportDataFilename), os.O_CREATE|os.O_WRONLY, ExportedFilesMode)
+	if err != nil {
+		return err
+	}
+	defer exportDataFile.Close()
+
+	swingStoreEntries := snapshotter.exporter.ExportSwingStore(ctx)
+
+	if len(swingStoreEntries) > 0 {
+		encoder := json.NewEncoder(exportDataFile)
+		encoder.SetEscapeHTML(false)
+		for _, dataEntry := range swingStoreEntries {
+			entry := []string{dataEntry.Path, dataEntry.Value}
+			err := encoder.Encode(entry)
+			if err != nil {
+				return err
+			}
+		}
+	}
```

Comment:
Could you add some commentary here explaining why the restoration process is reading from storage?